server/controllers/auth/signup.js

```diff
@@ -1,3 +1,39 @@
-exports.signup = (req, res) => {
-  res.send(' Hello from signup file');
+const { hash } = require('bcrypt');
+const { sign } = require('jsonwebtoken');
+const { signupSchema } = require('./validation/signup-validation');
+const { users: { getUserByUsername, getUserByEmail, addUser } } = require('../../database/queries');
+
+
+exports.signup = (req, res, next) => {
+  const key = process.env.KEY;
+  const {
+    username, email, password, nativeLangId, learningLangId,
+  } = req.body;
+
+  signupSchema.validate({
+    username, email, password, nativeLangId, learningLangId,
+  })
+    .then(() => getUserByUsername(username))
+    .then((result) => {
+      if (result.rows.length !== 0) throw Error('username exists');
+    })
+    .then(() => getUserByEmail(username))
+    .then(({ rows }) => {
+      if (rows.length !== 0) throw Error('email exists');
+    })
+    .then(() => hash(password, 10))
+    .then((hashed) => addUser({
+      username, email, password: hashed, nativeLangId, learningLangId,
+    }))
+    .then(() => getUserByUsername(username))
```

Comment:
in postgres, you can use the [RETURNING](https://stackoverflow.com/questions/2276574/returning-inserted-rows-in-postgresql) construct in an INSERT query (in `addUser()`), so that you can directly return the inserted row and don't have to do another roundtrip to the database to retrieve it.